Use AsyncStream<Never> for waitUnlessCancelled

KaiOelfke opened the pull request:
withUnsafeContinuation needs a withTaskCancellationHandler to support cooperative cancellation. One can use the continuation of an AsyncStream instead. AsyncStream supports cancellation. An element type of Never guarantees that there will be no emits. So the stream can only be cancelled or finished.

Credit for this goes to the @pointfreeco episode https://www.pointfree.co/episodes/ep210-clocks-controlling-time where this is approach is explained for the TestClock implementation.

See: https://github.com/pointfreeco/swift-clocks/blob/d8ff0c92e8c9d139356adae1c3c079b12e81d4df/Sources/Clocks/TestClock.swift#L117

groue commented:
Hello @KaiOelfke, thank you for this pull request.

What does it fix?

KaiOelfke commented:
> What problem does it fix?

There's no problem. This is not a PR to change behavior, but a refactoring. This approach in my opinion is simpler compared to the combination of withUnsafeContinuation and withTaskCancellationHandler. But if you think the current code is easier to reason about feel free to close this PR.

Here's a quote explaining why they choose AsyncStream.

> "The reason our clock is not cooperatively cancelling is because of the usage of withUnsafeContinuation:
>
> If the surrounding asynchronous context that this is running in is cancelled, it will continue suspending. this isn’t documented in Swift or in the evolution proposals, but it does seem to be on purpose.
>
> If you want to this piece of code to participate in cooperative cancellation, the easiest way to do so is to use an AsyncStream instead. In many ways their APIs are very similar, but streams do cooperatively cancel by breaking out of a for await when it detects cancellation.
>
> We’re even able to use Never as the output type of the stream since we don’t need it to ever emit. We just need it to exist until it is finished."

groue commented:
All right, I'm reconsidering your idea. You have deleted the explanation for the use of a recursive lock: is it because such a lock is no longer necessary? Shouldn't we change the lock type to a plain `NSLock`, then?

KaiOelfke commented:
> All right, I'm reconsidering your idea. You have deleted the explanation for the use of a recursive lock. 

This got lost, when deleting the withTaskCancellationHandler as the AsyncThrowingStream replaces it. I didn't think too much about using a NSLock or NSRecursiveLock, when making the initial change.

> Is it because such a lock is no longer necessary? Shouldn't we change the lock type to a plain `NSLock`, then?

But I think it's indeed no longer necessary to use a recursive lock so I changed it to NSLock now.

I added some more comments and one test as well. It's possible to test the cancellation behavior of AsyncThrowingStream with the tests in the other PR:

- `test_that_cancellation_while_suspended_increments_the_semaphore_two`
- `test_that_cancellation_before_suspension_increments_the_semaphore_two`
- `test_cancellation_while_suspended_throws_CancellationError`
- `test_cancellation_before_suspension_throws_CancellationError`

groue commented:
👍 I'm happy we no longer have to use an NSRecursiveLock. This opens the way to more modern locks eventually (`OSAllocatedUnfairLock`, or any other lock that looks relevant). Now I'm sold on this PR :-)

> It's possible to test the cancellation behavior of AsyncThrowingStream with the tests in the other PR

All right. I'll focus on #3 first!

groue commented:
Thank you @KaiOelfke, but I don't feel like this PR is necessary.
